java-checks/src/main/java/org/sonar/java/checks/SynchronizedClassUsageCheck.java

```diff
@@ -58,88 +59,83 @@
     .put("java.lang.StringBuffer", "\"StringBuilder\"")
     .put("java.util.Stack", "\"Deque\"")
     .build();
-  private JavaFileScannerContext context;
 
   @Override
-  public void scanFile(JavaFileScannerContext context) {
-    this.context = context;
-    scan(context.getTree());
+  public List<Tree.Kind> nodesToVisit() {
+    return ImmutableList.of(Tree.Kind.COMPILATION_UNIT);
   }
 
   @Override
-  public void visitVariable(VariableTree tree) {
-    super.visitVariable(tree);
-
-    boolean hasIssueOnDeclaredType = reportIssueIfDeprecatedType(tree.type());
-    if (!hasIssueOnDeclaredType) {
-      ExpressionTree init = tree.initializer();
-      if (init != null && init.is(Tree.Kind.NEW_CLASS)) {
-        reportIssueIfDeprecatedType(tree.initializer());
-      }
+  public void visitNode(Tree tree) {
+    if (!hasSemantic()) {
+      return;
     }
-  }
 
-  @Override
-  public void visitMethod(MethodTree tree) {
-    scan(tree.modifiers());
-    scan(tree.typeParameters());
-    scan(tree.returnType());
-    scan(tree.defaultValue());
-    scan(tree.block());
-    if (!isOverriding(tree)) {
-      for (VariableTree param : tree.parameters()) {
-        reportIssueIfDeprecatedType(param.type());
-      }
-      reportIssueIfDeprecatedType(tree.returnType());
+    DeprecatedTypeVisitor visitor = new DeprecatedTypeVisitor(this);
+    for (Tree typeTree : ((CompilationUnitTree) tree).types()) {
+      typeTree.accept(visitor);
     }
-
   }
 
-  @Override
-  public void visitClass(ClassTree tree) {
-    super.visitClass(tree);
-    for (TypeTree parent : tree.superInterfaces()) {
-      reportIssueIfDeprecatedType(parent);
-    }
+  private static class DeprecatedTypeVisitor extends BaseTreeVisitor {
 
-  }
+    private final IssuableSubscriptionVisitor check;
 
-  private boolean reportIssueIfDeprecatedType(@Nullable ExpressionTree tree) {
-    if (tree == null) {
-      return false;
+    public DeprecatedTypeVisitor(IssuableSubscriptionVisitor check) {
```

Comment:
You can remove the constructor and 'check' field by removing the 'static' modifier on the class and invoking addIssue directly (without prefixing it with check.) in reportIssueOnDeprecatedType.